rust/bindings/build.rs

```diff
@@ -0,0 +1,260 @@
+//! This crate interprets all '*.json' files in `/specs` as OpenRPC documents,
+//! and generates corresponding modules based on the filename.
+
+use anyhow::{bail, Context as _};
+use proc_macro2::TokenStream;
+use quote::quote;
+use schemars::schema::{
+    ArrayValidation, ObjectValidation, RootSchema, Schema, SchemaObject, SingleOrVec,
+    SubschemaValidation,
+};
+use syn::Ident;
+
+use std::{
+    env,
+    ffi::OsStr,
+    fs::{self, File},
+    path::Path,
+};
+
+fn main() -> anyhow::Result<()> {
+    let dir = concat!(env!("CARGO_MANIFEST_DIR"), "/specs/");
+    println!("cargo::rerun-if-changed={}", dir);
+
+    let mut modules = TokenStream::new();
+    for entry in fs::read_dir(dir)? {
+        let entry = entry?;
+        let path = entry.path();
+        if entry.metadata()?.is_file()
+            && path.extension().is_some_and(|it| it == "json")
+            && path
+                .file_name()
+                .is_some_and(|it| !it.to_string_lossy().starts_with('.'))
+        {
+            modules.extend(
+                file2module(&path).context(format!("couldn't process file {}", path.display()))?,
+            )
+        }
+    }
+
+    let pretty = prettyplease::unparse(&syn::parse2(modules).context("internal codegen error")?);
+    let mut out = env::var_os("OUT_DIR").context("invalid build script environment")?;
+    out.push("/bindings.rs");
+    fs::write(out, pretty).context("couldn't write generated code")?;
+
+    Ok(())
+}
+
+fn file2module(path: &Path) -> anyhow::Result<TokenStream> {
+    let file_stem = path
+        .file_stem()
+        .and_then(OsStr::to_str)
+        .context("bad or no file name")?;
+    let mod_name = syn::parse_str::<syn::Ident>(
+        file_stem
+            .split_once(char::is_whitespace)
+            .map(|(before, _after)| before)
+            .unwrap_or(file_stem),
+    )
+    .context("filename must of the form `IDENT` or `IDENT + WHITESPACE + IGNORED`")?;
```

Comment:
non-blocking nit: as much as I like to decipher such docs, it would be great to have 
```
"filename must of the form `IDENT` or `IDENT + WHITESPACE + IGNORED`, e.g. `some actual filename`"
```